src/segment/filter/blocked_bloom/mod.rs

```diff
@@ -0,0 +1,212 @@
+// Copyright (c) 2024-present, fjall-rs
+// This source code is licensed under both the Apache 2.0 and MIT License
+// (found in the LICENSE-* files in the repository)
+
+mod builder;
+use super::{bit_array::BitArrayReader, AMQFilter, CACHE_LINE_BYTES};
+use crate::{
+    coding::{DecodeError, Encode, EncodeError},
+    file::MAGIC_BYTES,
+};
+pub use builder::Builder;
+use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
+use std::io::{Read, Write};
+
+/// Two hashes that are used for double hashing
+pub type CompositeHash = (u64, u64);
+
+pub struct BlockedBloomFilter {
+    /// Raw bytes exposed as bit array
+    inner: BitArrayReader,
+
+    /// Number of hash functions
+    k: usize,
+
+    /// Number of blocks in the blocked bloom filter
+    num_blocks: usize,
+}
+
+impl AMQFilter for BlockedBloomFilter {
+    fn bytes(&self) -> &[u8] {
+        self.inner.bytes()
+    }
+
+    /// Size of bloom filter in bytes
+    #[must_use]
+    fn len(&self) -> usize {
+        self.inner.bytes().len()
+    }
+
+    /// Returns `true` if the hash may be contained.
+    ///
+    /// Will never have a false negative.
+    #[must_use]
+    fn contains_hash(&self, (mut h1, mut h2): CompositeHash) -> bool {
+        let block_idx = h1 % (self.num_blocks as u64);
+
+        for i in 1..(self.k as u64) {
+            h1 = h1.wrapping_add(h2);
+            h2 = h2.wrapping_add(i);
```

Comment:
This should probably be `wrapping_mul`, I already changed it in the standard filter; not sure how much of a difference it makes, but I think the enhanced double hashing formula is `h(1) + i * h2 + f(i)`

Same in Builder